feat(CI): check for badly formatted titles or missing/contradictory labels

grunweg opened the pull request:
---

The zulip workflow is entirely cargo-culted from #16296; please review carefully.

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

For details on the "pull request lifecycle" in mathlib, please see:
https://leanprover-community.github.io/contribute/index.html

In particular, note that most reviewers will only notice your PR
if it passes the continuous integration checks.
Please ask for help on https://leanprover.zulipchat.com if needed.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

If you are moving or deleting declarations, please include these lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Moves:
- Vector.* -> Mathlib.Vector.*
- ...

Deletions:
- Nat.bit1_add_bit1
- ...

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]

-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

grunweg committed:
wip: add ci workflow; untested yet

bryangingechen commented on .github/workflows/build.yml.in:
```diff
@@ -58,6 +58,17 @@ jobs:
+        run: |
+          set -eo pipefail
+          title="${{github.event.pull_request.title}}"
+          echo $title
```
This needs to be quoted otherwise `*` etc in the title might be interpreted as a glob
```suggestion
          echo "$title"
```

bryangingechen commented on .github/workflows/build.yml.in:
```diff
@@ -58,6 +58,17 @@ jobs:
+          title="${{github.event.pull_request.title}}"
+          echo $title
+          label_names=$(echo "${{ toJson(github.event.pull_request.labels) }}" | jq -r '.[].name')
+          echo $label_names
```
Same as above
```suggestion
          echo "$label_names"
```

bryangingechen commented on .github/workflows/build.yml.in:
```diff
@@ -58,6 +58,17 @@ jobs:
+        shell: bash # just in case
+        run: |
+          set -eo pipefail
+          title="${{github.event.pull_request.title}}"
```
I think Actionlint suggests using `env` in the workflow step, cf. [the example here](https://docs.github.com/en/actions/security-for-github-actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable)).

bryangingechen commented on .github/workflows/build.yml.in:
```diff
@@ -61,13 +61,14 @@ jobs:
           label_names=$(echo "${{ toJson(github.event.pull_request.labels) }}" | jq -r '.[].name')
           echo "$label_names"
-          lake exe check-title-labels $title $label_names
+          lake exe check-title-labels $TITLE $label_names
```
Ah, now that I actually try to parse the position indicators in the error messages, it looks like there are still quoting issues in this line:
```suggestion
          lake exe check-title-labels "$TITLE" "$label_names"
```

grunweg committed:
Update .github/workflows/build.yml.in

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

grunweg commented:
Thanks again!

grunweg committed:
Synchronise workflows

joneugster commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,106 @@
+    return true
+  else if normalised_labels.contains "WIP" && ["awaiting-review", "bors"].any normalised_labels.contains then
+    return true
+  else if ["awaiting-author", "awaiting-zulip"].all normalised_labels.contains then
+    return true
+  else if ["bors", "WIP"].all normalised_labels.contains then
+    return true
```
Aren't these already covered above?

joneugster commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,106 @@
+
+
+open Cli in
+/-- Implementation of the `mk_all` command line program.
+The exit code is the number of files that the command updates/creates. -/
```
update

grunweg replied:
Indeed; done

joneugster commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,106 @@
+
+  ARGS:
+    title : String; "this PR's title"
+    labels : String; "space-separated list of label names of this PR"
```
```suggestion
    ...labels : String; "Label name(s) of this PR"
```
I think this should give you an `Array String` of labels which you can access with

```lean
let labels : Array String := args.variableArgsAs! String
```

so you don't need to split manually.

See `outputs` in [import-graph](https://github.com/leanprover-community/import-graph/blob/0a756018b736ef4f4bc3c9a45fc03cf97a0b1a0c/Main.lean#L28-L33) for an example.

grunweg replied:
Ah, that's the syntax for passing an array of arguments. Much nicer indeed, thank you!

joneugster commented on .github/workflows/build.yml.in:
```diff
@@ -58,6 +58,18 @@ jobs:
           lake exe lint-style
 
+      - name: "check the formatting of the PR title and for missing or inconsistent labels"
+        # TODO: skip this check if the PR is in draft stage!
```
```suggestion
        if: github.event.pull_request.draft == false
```
should do this

grunweg committed:
Skip if the PR is in draft stage

Co-authored-by: Jon Eugster <eugster.jon@gmail.com>

grunweg commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,106 @@
+  if !known_kinds.contains (parts.get! 0) then
+    IO.println s!"error: the PR kind should be one of {", ".intercalate (known_kinds.map (s!"\"{·}\""))}"
+    numberErrors := numberErrors + 1
+  return numberErrors
```
- I added checks for scope being well-formed
- checking the directory exists on disk is not ideal. I *could* read in `Mathlib.lean` (this is what `lint-style.lean` does) --- do you mind if I do so in a follow-up?

grunweg reviewed (commented):
Thanks for your careful review! I just revamped the logic, which should be much more robust now.

bryangingechen commented on .github/build.in.yml:
```diff
@@ -63,6 +63,18 @@ jobs:
           lake exe lint-style
 
+      - name: "check the formatting of the PR title and for missing or inconsistent labels"
+        if: github.event.pull_request.draft == false
```
```suggestion
        if: github.event.pull_request.draft == false and github.event.pull_request.base == 'master'
```
It might be good to add a check that the PR is into `master` since sometimes people open PRs into other PR branches and presumably we don't care about the title or labels for those.

grunweg replied:
Good point, added!

grunweg committed:
feat: add a script to check for PR title and labelling

The script checks most items from the commit convention,
ensures that a feature PR has a topic/area label,
and errors if any two labels are contradictory.

Add detailed tests for the title validation.

grunweg committed:
Does this help?

grunweg committed:
Synchronise

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

grunweg commented:
CI looks green now: I don't mind if this is reviewed together with #17649 or separately.

bryangingechen commented:
I don't feel qualified to review the Lean code - @joneugster does everything look good to you now?

Command-Master commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,267 @@
+  else if main.containsSubstr "  " then
+    errors := errors.push "error: the PR title contains multiple consecutive spaces; please add just one"
+  let main := main.removeLeadingSpaces
+  if main.front.toLower != main.front then
+    errors := errors.push "error: the main PR title should be lowercased"
```
This could give false positives on acronyms

Command-Master commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,267 @@
+      errors := errors.push s!"error: the PR scope must not start with 'Mathlib/'"
+    if scope.endsWith ".lean" then
+      errors := errors.push s!"error: a PR's scope must not end with '.lean'"
+    -- Future: we could further validate the scope, that this is a valid module or directory
```
Perhaps also check whether scope is `/` separated or `.` separated?

grunweg replied:
Good idea, done.

Command-Master commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,267 @@
+  else if normalised_labels.contains "bors" &&
+      ["awaiting-CI", "awaiting-zulip", "WIP"].any normalised_labels.contains then
+    return true
+  else if normalised_labels.contains "awaiting-zulip" &&
+      ["delegated", "awaiting-zulip"].all normalised_labels.contains then
```
The check for "awaiting-zulip" is duplicated here

Command-Master reviewed (commented):
Left a few comments. I think overall it looks great, but a way to suppress it in case of false positives would be nice to have

Command-Master commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,267 @@
+      errors := errors.push s!"error: the PR type should be of the form abbrev(scope)"
+    let idx := type.find (· == '(')
+    let idx2 := type.find (· == ')')
+    if idx2 < idx then
+      errors := errors.push "error: mismatched parentheses; the PR type should be of the form abbrev(scope)"
```
This should be covered by checking that the type ends with a `)`

grunweg replied:
It mostly was, but not quite. I've added a few more tests --- but thanks for pointing me in this direction!

adomani commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,267 @@
+#guard_msgs in
+#check_title "feat(test) : (confusing) bad title"
+
+/-- info: Message: 'error: the main PR title should be lowercased' -/
+#guard_msgs in
+#check_title "feat: My Bad Title"
+
```
```suggestion
```
This test seems to be a duplicate of a previous one.

adomani reviewed (commented):
Thanks for writing this!

I had a quick read through and I wonder whether some of the checks are a little too restrictive (e.g., the one about lower-casing the title).

Could you run the `#check_title` script on a sample of actual PR titles to verify that it only flags titles that do indeed look malformed in some way?

Moreover, is this check triggered only by push?  Does this mean that the only way to get CI to pass is to change the title and then push a change to the PR?

Maybe this check should take place in a separate CI workflow and be triggered by modification of the title (assuming that this is possible, I'm not sure).

Finally, it may be easier to start with a single check and extend the scope of the validation of the title step by step.

grunweg committed:
Address most review comments, make tests more comprehensive.

grunweg committed:
Allow upper-cased titles for now.

grunweg commented:
Thanks for the careful review comments! I just pushed an update, taking them all into account.
- added a few more tests
- I've disabled the check for upper-case titles for now. I don't think it's as important.

I'll run this through the current PR titles next and see which ones are flagged.

grunweg committed:
Reduce checks slightly further.

grunweg commented:
Running the check in all open PRs' titles yields 176 violations --- but also counting PRs marked as `WIP`, on which this check won't run. I've disabled one more check, bringing that number to 153. All of them look technically legit.

grunweg committed:
Merge branch 'master' into MR-bad-titles

grunweg committed:
Document the new script.

edegeltje commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,274 @@
+    let titleErrors := validateTitle title
+    for err in titleErrors do
+      IO.println err
+      numberErrors := numberErrors + 1
+    -- TODO: can I use this, with some casting?
+    -- numberErrors := numberErrors + titleErrors.size
```
you can use [Nat.toUInt32](https://leanprover-community.github.io/mathlib4_docs/Init/Data/UInt/BasicAux.html#Nat.toUInt32) (or maybe preferably [UInt32.ofNatTruncate](https://leanprover-community.github.io/mathlib4_docs/Init/Data/UInt/BasicAux.html#UInt32.ofNatTruncate)?) for that. This cast will likely never be implicit, since it changes data.

edegeltje commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,274 @@
+    -- A feature PR should have a topic label.
+    if title.startsWith "feat" && !labels.any
+        (fun s ↦ s.startsWith "t-" || ["CI", "IMO"].contains s) then
+      IO.println "error: feature PRs should have a 't-something' or the 'CI' label"
```
```suggestion
      IO.println "error: feature PRs should have a 't-something',  'CI' or 'IMO' label"
```

edegeltje commented on .github/build.in.yml:
```diff
@@ -63,6 +63,18 @@ jobs:
         run: |
           lake exe lint-style
 
+      - name: "check the formatting of the PR title and for missing or inconsistent labels"
+        if: github.event.pull_request.draft == false && github.event.pull_request.base == 'master'
+        shell: bash
+        env:
+          TITLE: ${{ github.event.pull_request.title }}
+        run: |
+          set -eo pipefail
+          echo "$TITLE"
+          label_names=$(echo "${{ toJson(github.event.pull_request.labels) }}" | jq -r '.[].name')
+          echo "$label_names"
+          lake exe check-title-labels "$TITLE" "$label_names"
+
```
do we want to have this run every time a commit gets pushed? or maybe only on creating the pr and on editing the pr title?

edegeltje commented on scripts/check-title-labels.lean:
```diff
@@ -0,0 +1,274 @@
+/-- The entrypoint to the `lake exe check-title-labels` command. -/
+def main (args : List String) : IO UInt32 := checkTitleLabels.validate args
+
+-- Tests for the PR title validation logic.
```
does it make sense to have a dedicated test file for this?

edegeltje reviewed (commented):
thanks for writing this!

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.

grunweg committed:
Merge branch 'master' into MR-bad-titles

grunweg committed:
Try workflow again; please review carefully

grunweg committed:
chore: split into three files

grunweg committed:
Apply suggestions from code review

Co-authored-by: blizzard_inc <edwardvdmeent@gmail.com>

grunweg committed:
Better cast

grunweg committed:
Reorder
